proof-systems/src/darlin/fiat-shamir/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+    // the concatenated byte sequence of elements.
+    seed_bytes: Vec<u8>,
+}
+
+impl FiatShamirRngSeed {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_bytes<'a, T: 'a + CanonicalSerialize>(&mut self, elem: &'a T) -> Result<&mut Self, Error> {
+        // Check we have not reached the maximum allowed seed size
+        if self.num_elements == u64::MAX {
+            return Err(Error::BadFiatShamirInitialization(format!(
+                "Maximum seed length {} exceeded",
+                u64::MAX
+            )));
+        }
+
+        // Get elem bytes and check that they are not over the maximum allowed elem len
+        let mut elem_bytes = serialize_no_metadata!(elem).map_err(|_| {
+            Error::BadFiatShamirInitialization("Unable to convert elem to bytes".to_owned())
+        })?;
+        let elem_bytes_len: u64 = elem_bytes.len().try_into().map_err(|_| {
+            Error::BadFiatShamirInitialization(format!(
+                "Max elem length exceeded. Max: {}",
+                u64::MAX
+            ))
+        })?;
+
+        // Update internal state
+        self.num_elements += 1;
+        self.elements_len.push(elem_bytes_len);
+        self.seed_bytes.append(&mut elem_bytes);
+        Ok(self)
+    }
+
+    pub fn finalize(self) -> Result<Vec<u8>, Error> 
+    {
+        serialize_no_metadata![self.num_elements, self.elements_len, self.seed_bytes]
+            .map_err(|e| {
+                Error::BadFiatShamirInitialization(format!("Unable to finalize seed: {:?}", e))
+        })
+    }
+}
+
+/// Thin wrapper around a type which is ToConstraintField + CanonicalSerialize
+pub trait Absorbable<F: Field>: ToConstraintField<F> + CanonicalSerialize {}
+
+impl<F: Field, T: ToConstraintField<F> + CanonicalSerialize> Absorbable<F> for T {}
+
+/// General trait for Fiat-Shamir transform, designed as a Sponge-based construction.
+pub trait FiatShamirRng: Sized + Default {
+    /// Internal State
+    type State: Clone + Debug;
+
+    /// Create a new `Self` by initializing its internal state with a fresh `seed`.
+    fn from_seed(seed: Vec<u8>) -> Result<Self, Error>;
+
+    /// Refresh the internal state with new material
+    fn absorb<F: Field, A: Absorbable<F>>(&mut self, to_absorb: A) -> Result<&mut Self, Error>;
+
+    /// Squeeze a new random field element, changing the internal state.
+    fn squeeze<F: PrimeField>(&mut self) -> Result<F, Error> {
+        Ok(self.squeeze_many(1)?[0])
+    }
+
+    /// Squeeze 'num' many random field elements, changing the internal state.
+    /// Depending on the internal implementation, it might be more
+    /// efficient than calling 'squeeze()' num times.
+    fn squeeze_many<F: PrimeField>(&mut self, num: usize) -> Result<Vec<F>, Error>;
+
+    /// Squeeze a new random field element having bit length of 128, changing the internal state.
+    /// NOTE: We require the G: EndoMulCurve generic for backward compatibility reasons: we don't
```

Comment:
The comments documenting the squeeze functions declare that their invocation does change the sponge internal state. Thus a user would be led to expect that inserting a squeeze in any place of an absorb-squeeze sequence would change the subsequent squeezed elements. However the current implementers of the `FiatShamirRng` trait, namely `FiatShamirChaChaRng` and `PoseidonSponge`, do not exhibit this behavior.

As an example, the last squeezed element obtained by performing the sequence
`init(seed); absorb(); absorb(); squeeze(); absorb(); squeeze()`
is the same as the element obtained by performing
`init(seed); absorb(); absorb(); absorb(); squeeze()`

This is true both for the `FiatShamirChaChaRng` and `PoseidonSponge` implementations. As discussed offline, we should understand if this behavior is consistent with the common understanding of a cryptographic sponge

- if yes, then the comments should be clarified
- if not, then the implementations of `FiatShamirChaChaRng` and `PoseidonSponge` should be changed

Reply:
Just to be clearer here, as the main focus was to minimize the number of permutations:

- The fact that `absorb()` doesn't always trigger a permutation seems to be common at different implementations I saw (in fact some of them, maybe to not contradict the specifications, calls this function in a different way and doesn't expose a public `absorb()`);
- When squeezing after absorbing, I will return the output without triggering any permutation if that was just performed. This seems the main points that contradicts the specifications, as it seems that the `squeeze()` should always trigger a state change (a permutation in our case);
- One optimization that is not adopted here, but that seems to be used in a lot of places, is to use up to RATE elements from the state as the output of `squeeze()`: currently, if I need to squeeze n challenges, I will trigger up to n permutations, while with this optimization they will become only n/RATE. At first glance it doesn't seem sound, as the squeeze outputs are supposed to be independent from each other, but maybe this is not true

Reply:
@DDT92

Reply:
@DanieleDiBenedetto 
In the [Sponge paper](https://keccak.team/files/SpongeFunctions.pdf) both the absorb() and squeeze() function run a permutation.
In our current implementation absorb() always runs the permutation, while squeeze does it only if there were no previous squeezes. This makes some sense in my opinion:
The squeezing is just copying a state element, which should be "random". Obviously if you run two consecutive squeeze() then you will add to the output two consecutive equal elements, otherwise the element is already "random" because previously you ran absorb().
About your other points:

- How do they do precisely in other implementations? Do they run a bunch consecutive absorb() and then a permutation after them? This case still makes sense, since the next output element will be "random" anyway.
- Where did you see the specs?
- It is still sound in my opinion: in fact it is like if - with reference to [Sponge paper](https://keccak.team/files/SpongeFunctions.pdf) - you are choosing  a bigger "alphabet group" (a consequently you are squeezing a "bigger" value) but the protocol remains the same in structure. In fact, all the element of the states are "random".

I think that there is the usual trade-off security-efficiency. Maybe it is possible to use all the optimizations and still being secure enough, but it would need a deeper security analysis.